src/session_state.rs

```diff
@@ -1,30 +1,34 @@
 /// This module represents the session state of an MQTT communication session.
 ///
 ///
+use crate::QoS;
 use embedded_nal::IpAddr;
-use heapless::{String, Vec};
+use heapless::{String, Vec, LinearMap};
 
 use embedded_time::{
     duration::{Extensions, Seconds},
-    Clock, Instant,
+    Clock as HwClock,
+    Instant,
 };
 
-pub struct SessionState<C: Clock> {
+pub struct SessionState<Clock: HwClock, const MSG_SIZE: usize, const MSG_COUNT: usize>
+{
     // Indicates that we are connected to a broker.
     pub connected: bool,
     pub keep_alive_interval: Option<Seconds<u32>>,
-    pub ping_timeout: Option<Instant<C>>,
+    pub ping_timeout: Option<Instant<Clock>>,
     pub broker: IpAddr,
     pub maximum_packet_size: Option<u32>,
     pub client_id: String<64>,
-    last_transmission: Option<Instant<C>>,
+    last_transmission: Option<Instant<Clock>>,
     pub pending_subscriptions: Vec<u16, 32>,
+    pub pending_publish: LinearMap<u16, Vec<u8, MSG_SIZE>, MSG_COUNT>,
```

Comment:
We should check that even if we delete messages from the middle the ordering is always based on when the message has been added to satisfy:
> When the Client re-sends any PUBLISH packets, it MUST re-send them in the order in which the original PUBLISH packets were sent (this applies to QoS 1 and QoS 2 messages)
This has also ramifications for implementation of QoS 2 since the ordering should be kept across the QoS options.

Reply:
The struct uses swap_remove for removing which I assume means putting the last value in the backing vector in the removed item's position.

The structure should be tweaked to not reorder or another container has to be found